A HACKING.adoc file with hacking advice

gasche opened the pull request:
I think that the README should present the compiler distribution, but does need to contain information that is only relevant to people hacking on it, currently it's trying to do a bit of both. There are various things on how to effectively hack that are folklore knowledge and we could profitably store in this file.

Cc @johnwhitington, our precious README steward.

Drup committed:
Starting Hacking.adoc.

gasche committed:
first draft of HACKING.adoc file [WIP]

Drup committed:
Backport gasche's improvements.

Drup committed:
Fix link.

Drup committed:
Extend the typecheking part.

Drup committed:
Various small fixes/improvements.

gasche committed:
add explanations on INRIA's CI server and how to use ocaml/precheck

gasche committed:
add a Changes entry

Drup commented:
Hum, https://github.com/ocaml/ocaml/pull/751  ?

gasche commented:
I guess I didn't know or forgot about it, sorry. (The fact that I didn't reply to your PR suggest that it was made in the time window where I disabled notification for new PRs in a desperate and ultimately doomed attempt to spend less time doing OCaml stuff.)

My HACKING.adoc file was sitting in a branch somewhere on my computer and I wanted to have it today as I was using the `precheck` thing. I'm sure I discussed it on some mantis ticket somewhere but I couldn't find it again.

Can we merge our files? Are you volunteering it do it, or should I add it to my TODO list?

Drup commented:
I remember taking some bit and pieces of your `Hacking.adoc`, but it was several months ago. I can take another look, but I think it's best if you just add your stuff on top of it (so that you can decide yourself how to improve things).

gasche committed:
merge all of Drup's Hacking.adoc content into HACKING.adoc

gasche committed:
split HACKING.adoc with parts in {parsing,typing}/HACKING.adoc

In the interest of keeping HACKING.adoc not-too-long and
general-purpose, advice on modifying specific sub-systems of the
compiler distribution should be moved to the directory of this
sub-system.

This PR also clarifies the relations between the README, INSTALL,
HACKING and CONTRIBUTING documents.

gasche commented:
@Drup I merged your branch and made more changes. What do you think?

dra27 commented on CONTRIBUTING.md:
```diff
@@ -7,9 +7,17 @@ OCaml distribution. These are just guidelines, not rules, use your
 in a pull request.
 
+This document assumes that you already have a patch against the
+compiler distribution sources that you wish to submit to the OCaml
```
`distribution compiler sources`?

dra27 commented on CONTRIBUTING.md:
```diff
@@ -7,9 +7,17 @@ OCaml distribution. These are just guidelines, not rules, use your
+compiler distribution sources that you wish to submit to the OCaml
+maintainers upstream. If you are looking for a document on how to
+build the compiler distribution from sources and install it, see
+[INSTALL.adoc](INSTALL.adoc) instead. If you are looking for
```
`See [INSTALL.adoc](INSTALL.adoc) for details on how to build the compiler
 distribution from sources.` (shorter sentence)

dra27 commented on CONTRIBUTING.md:
```diff
@@ -7,9 +7,17 @@ OCaml distribution. These are just guidelines, not rules, use your
+build the compiler distribution from sources and install it, see
+[INSTALL.adoc](INSTALL.adoc) instead. If you are looking for
+a document on how to hack on the compiler distribution sources, see
+[HACKING.adoc](HACKING.adoc) instead.
```
`See [HACKING.adoc](HACKING.adoc) for details on how to hack the compiler
 distribution sources.` (shorter sentence; remove word "on").

Possibly "edit" rather than "hack"?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+2. Consult link:INSTALL.adoc[] for build instructions. Here is the gist of it:
++
+----
+opam compiler-conf configure
```
This step is presumably in some way optional - can that be mentioned?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+* The
+  https://github.com/ocamllabs/compiler-hacking/wiki/Things-to-work-on[OCaml
+  Labs compiler-hacking wiki] contains various ideas of changes to
+  propose, some easy, some requiring sensibly more work.
```
sensibly → realistically? Either should be before "requiring" (i.e. "realistically requiring more work")?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  Labs compiler-hacking wiki] contains various ideas of changes to
+  propose, some easy, some requiring sensibly more work.
+
+* Documentation improvements are always most appreciated, either in
```
"much appreciated" :smile:

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  propose, some easy, some requiring sensibly more work.
+
+* Documentation improvements are always most appreciated, either in
+  the various mli files or in the official manual
```
".mli files" (i.e. with the dot) or "header files"?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  the various mli files or in the official manual
+  (See link:manual/README.md[]). If you invest effort in understanding
+  a part of the codebase, submitting a pull request that adds
+  clarifying comment can be an excellent contribution to help future
```
comment → comments (or "adds a clarifying comment").

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  (See link:manual/README.md[]). If you invest effort in understanding
+  a part of the codebase, submitting a pull request that adds
+  clarifying comment can be an excellent contribution to help future
+  you and your peers.
```
"to help you and your peers in future." (it's a slightly condescending way of putting it, but I can't think of a better phrase at the moment)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  various parts of the compiler from the angle of a specific aspect or
+  feature.
+
+   Again, reviewing small or medium-sized pull requests is accessible
```
Ultra-pedantic: stray space at the start of the line?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+   Again, reviewing small or medium-sized pull requests is accessible
+  to anyone with OCaml programming experience, and helps maintainers
+  and other contributors. If you also submit pull requets yourself,
```
Typo: requets → requests

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+   Again, reviewing small or medium-sized pull requests is accessible
+  to anyone with OCaml programming experience, and helps maintainers
+  and other contributors. If you also submit pull requets yourself,
+  a good discipline to follow is to review at least as many pull
```
Delete "to follow".

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+== Structure of the compiler
+
+The compiler code base can be intimidating at first sight. Here are
```
"codebase" (one word elsewhere)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+compilation. It is in charge of parsing the command-line arguments,
+and then composing the compiler passes according to the invoked
+command-line, by calling functions from the various parts of the
+compiler described below.
```
Reword "command-line arguments, ..." to "command-line arguments to determine the sequence of compiler passes which are invoked by calling functions from the various parts of the compiler described below."?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+(link:parsing/parsetree.mli[] has lot of helpful comments). See
+link:parsing/HACKING.adoc[].
+
+The logic for Camlp4 and Ppx preprocessing is not in link:parsing/[],
```
Is there Camlp4 stuff still anywhere in the compiler?

gasche replied:
the `-pp` option still exists, although its logic does not depend on camlp4 in particular. It takes the source file in input, as opposed to the parsed AST like `-ppx`.

dra27 replied:
Ah, I see - might `The logic for classic preprocessing (-pp option; camlp4, etc.) and AST preprocessing (-ppx option) is not in..." be better? (AST preprocessing is a term I just made up... what's the proper heading for -ppx??)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+==== Typing -- link:typing/[]
+
+Type-checks syntactic AST and produces a typed representation
```
Terminologically (!!), it seems strange to me to refer to a syntactic Abstract Syntax Tree. Is a word needed at all in this context?

`typed representation` → `typed version` ?

gasche replied:
I prefer "representation": it's not a different version, it's a different view of the same program.

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+=== Libraries
+
+link:stdlib/[]:: The standard library. Each file is mostly independent and should not need further knowledge.
```
mostly → largely

`independent` → `standalone` or "independent of the others"? The "and ..." feels superfluous to me.

Is it worth mentioning the internal modules (especially the modules related to Format, etc.) straightaway?

gasche replied:
> Is it worth mentioning the internal modules (especially the modules related to Format, etc.) straightaway?

This would be interesting but should go in `stdlib/HACKING.adoc` I think, in the interest of keeping the central file concise and focused on generic, subsystem-agnostic information advice.

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  INSTALL.adoc::          instructions for installation
+  LICENSE::               license and copyright notice
+  Makefile::              main Makefile
+  Makefile.nt::           MS Windows Makefile
```
Nit-pick (which I think I consistently adhered to when updating the Windows instructions). MS-DOS is a specific trademark - the product is correctly named Microsoft Windows (or Microsoft® Windows)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  Makefile.shared::       common Makefile
+  Makefile.tools::        used by manual/ and testsuite/ Makefiles
+  README.adoc::           this file
+  README.win32.adoc::     info on the MS Windows ports of OCaml
```
MS -> Microsoft® (as above)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  Makefile.nt::           MS Windows Makefile
+  Makefile.shared::       common Makefile
+  Makefile.tools::        used by manual/ and testsuite/ Makefiles
+  README.adoc::           this file
```
No it's not!

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  Changes::               what's new with each release
+  configure::             configure script
+  CONTRIBUTING.md::       how to contribute to OCaml
+  HACKING.adoc::          tips for people modifying the code
```
"this file"

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  bytecomp/::             bytecode compiler and linker
+  byterun/::              bytecode interpreter and runtime system
+  compilerlibs/::         the OCaml compiler as a library
+  config/::               autoconfiguration stuff
```
"configuration files" (it's not automatic for Windows!)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  driver/::               driver code for the compilers
+  emacs/::                editing mode and debugger interface for GNU Emacs
+  experimental/::         experiments not built by default
+  flexdll/::              empty (see README.win32.adoc)
```
Git submodule (initialise and populate using `git submodule update --init`)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+  manual/::               system to generate the manual
+  middle_end/::           the flambda optimisation phase
+  ocamldoc/::             documentation generator
+  otherlibs/::            several external libraries
```
"external" doesn't feel right here, given that the threads implementations are very much tied to the standard library. Perhaps "optional"?

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+=== Bootstrapping
+
+The OCaml compiler is bootstrapped. This means that a previous version of the OCaml compiler (along with various tools and a compiled version of the standard library) is included in the repository under the link:boot/[] directory.
```
boot/ doesn't include the standard library _in the repository_. How about `This means that previously-compiled bytecode versions of the compiler, dependency generator and lexer are included in the repository under the link:boot/[] directory. These bytecode images are used once the bytecode runtime (which is written in C) has been built to compile the standard library and then to build a fresh compiler.`

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+
+INRIA provides a Jenkins continuous integration service that OCaml
+uses, see link:https://ci.inria.fr/ocaml/[]. It provides a wider
+architecture support (MSVC and MingW, a zsystems s390x machine, and
```
"MinGW" (note case)

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+INRIA provides a Jenkins continuous integration service that OCaml
+uses, see link:https://ci.inria.fr/ocaml/[]. It provides a wider
+architecture support (MSVC and MingW, a zsystems s390x machine, and
+various MacOS versions) than the Travis/AppVeyor testing on github,
```
Personally, I'd put the bit in brackets after the word GitHub - i.e. "It provides a wider...testing on GitHub (MSVC and MinGW, ...)."

gasche replied:
I find the position you suggest ambiguous, as the location could suggest that we are listing what is available on github. I'll leave it as is for now -- hopefully it will be even clearer once a short paragraph on Github's CI has been written above.

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+uses, see link:https://ci.inria.fr/ocaml/[]. It provides a wider
+architecture support (MSVC and MingW, a zsystems s390x machine, and
+various MacOS versions) than the Travis/AppVeyor testing on github,
+but only runs on commits to the trunk or release branch, not on every
```
branch → branches
"commits **on** the..."?

gasche replied:
"commit to" seems equally common in this context, and I already use "on" in "runs on" so I'm tempted to leave as is.

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+but only runs on commits to the trunk or release branch, not on every
+PR.
+
+You do not need to be an INRIA employee to open account on this travis
```
Travis? Jenkins, isn't it?
"open **an** account"

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+You do not need to be an INRIA employee to open account on this travis
+service; anyone can create an account there to access build logs,
+enable email notifications, and manually restart builds. If you would
+like to do this but have trouble doing it, you may contact Damien
```
"you may contact" → "please contact"

dra27 commented on HACKING.adoc:
```diff
@@ -0,0 +1,254 @@
+If you have suspicions that a PR may fail on exotic architectures
+(it touches the build system or the backend code generator,
+for example) and would like to get wider testing than github's CI
+provides, it is possible to manually start INRIA's CI on arbitrary git
```
Nit-pick: here and in other places, isn't it meant to be written Inria? (also a few instances of GitHub not being capitalised)

gasche replied:
It's complicated, but let's say for now I like the historic spelling "INRIA" better -- it's also consistently used in Copyright headers etc.

dra27 replied:
Fair enough :)

dra27 commented on parsing/HACKING.adoc:
```diff
@@ -0,0 +1,9 @@
+
+link:parsing/location.mli[Location]:: This module contains utilities
+related to locations and error handling. In particular, it contains
+handler that are used for all the error reporting in the compiler.
```
handler → handlers

dra27 commented on parsing/HACKING.adoc:
```diff
@@ -0,0 +1,9 @@
+link:parsing/parsetree.mli[Parsetree] and link:parsing/asttypes.mli[Asttypes]::
+Parsetree is an AST of the surface language of OCaml. It is well
```
"surface language of OCaml"? Is there any reason not to sau "AST of the OCaml language"?

gasche replied:
I think Drup's intent was to separate "surface language" from other representations that have been simplified (for example syntactic sugar was expanded away), eg. the typedtree or the lambda-code, which represent distinct intermediate languages. I'll reformulate as "Parsetree is an Abstract Syntax Tree (AST) representation of OCaml source code".

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+
+http://caml.inria.fr/pub/docs/u3-ocaml/index.html[Using, Understanding, and Unraveling the OCaml Language by Didier Rémy] ::
+This book provides (among other things) a formal description of parts
+of the core OCaml language, starting by a simple Core ML.
```
"starting **with**"

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+understand the type-checker codebase, in particular its handling of
+polymorphism/generalization.
+
+After that, the best is to dive right in. There is no real "entry
```
"After that, it's best to dive..."

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+polymorphism/generalization.
+
+After that, the best is to dive right in. There is no real "entry
+point", but understanding of both the parsetree and the typedtree is
```
"but **an** understanding"

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+The datastructures ::
+link:typing/types.mli[Types] and link:typing/typedtree.mli[Typedtree]
+are the two main datastructures in the typechecker. They correspond to
+the surface language annotated with all the information needed for
```
Again, "surface language"?

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+the surface language annotated with all the information needed for
+type checking and type inference. link:typing/env.mli[Env] contains
+all the environments that are used in the typechecker. Each node in
+the typedtree is annotated with the local environment.
```
"with **a** local environment" (or "its")?

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+for a different part of the language:
+link:typing/typetexp.mli[Typetexp] for type expressions,
+link:typing/typecore.mli[Typecore] for the core language,
+link:typing/typecore.mli[Typemod] for modules,
```
Link is wrong - should be to typemod.mli

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+
+Note on dependencies between modules::
+Most of the modules presented above are inter-dependent with each
+other. Since OCaml prevents circular dependencies between files, the
```
Delete "with each other" (it's tautologous).
prevents → "doesn't permit"

dra27 replied:
prevents → "doesn't permit"

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+and finally link:typing/includeclass.mli[Includeclass] for the object
+system.
+
+Note on dependencies between modules::
```
Personally, I'd just put "Dependencies between modules"

dra27 commented on typing/HACKING.adoc:
```diff
@@ -0,0 +1,57 @@
+implementation uses forward declarations, implemented with references
+to functions that are filled later on. An example can be seen in
+link:typing/typecore.mli[Typecore.type_module], which is filled in
+link:typing/typecore.mli[Typemod].
```
Shouldn't the second one be `typemod.mli`?

gasche replied:
Right, and we should rather link to the `.ml` files here.

dra27 reviewed (commented):
Hope some of this is helpful - up to you if you'd prefer to change it here or in a subsequent GPR!

gasche committed:
HACKING.adoc: plenty of changes suggested by David Allsopp's review

gasche commented:
Thanks a lot @dra27 for the review. I acted on your comments.

I think I'll wait a bit more in case anyone has feedback, but then I'll go ahead and merge because (1) the stuff in there is already useful so why wait and (2) contributions of new content to these documents are best made by creating new PRs on top of the current state.

Drup commented on HACKING.adoc:
```diff
@@ -0,0 +1,259 @@
+  to anyone with OCaml programming experience, and helps maintainers
+  and other contributors. If you also submit pull requests yourself,
+  a good discipline is to review at least as many pull requests as you
+  submit.
```
The rendering of this paragraph is wrong (adoc interprets the indentation as a quotation)

gasche replied:
Thanks, fixed.

Drup commented on HACKING.adoc:
```diff
@@ -0,0 +1,259 @@
+  toplevel/::             interactive system
+  typing/::               typechecking
+  utils/::                utility libraries
+  yacc/::                 parser generator
```
I think we should not list all the files/directories. Some of them are duplicated with the sections above, and some of them are really not that interesting. In the end, it's just mostly noise.

gasche replied:
It was present in the initial README -- it comes from the very first version of this README -- and I think it's an amusing exercise. I left the list of files. It is more useful for the "small files" (which are not justified anywhere else) than for the directories indeed, so we could think of removing them, but let's not overdo it.

Drup reviewed (commented):
[Rendered version here](https://github.com/gasche/ocaml/blob/HACKING.adoc/HACKING.adoc).

Looks good to me.

gasche commented:
I merged.
